File: external/iup/src/emscripten/iupemscripten_label.js
/*global document alert IupCommon UTF8ToString IupCommon autoAddDeps mergeInto LibraryManager*/

var LibraryIupLabel = {
  //	$Button__deps: ['$CommonGlobals'],
  $Label: {
  },

  emjsLabel_CreateLabel: function() {
    // TODO: need to refactor to support image label
    // should create canvas element from image data in iup
    // also need one for separators (horizontal and vertical)
    var widget_object;
    widget_object = document.createElement("div");

    var handle_id = IupCommon.RegisterNewObject(widget_object);
    IupCommon.InitializeObject(widget_object);

    return handle_id;
  },

  emjsLabel_CreateImage: function() {
    var widget_object;
    //widget_object = document.createElement("CANVAS");
    widget_object = document.createElement("IMG");
    var handle_id = IupCommon.RegisterNewObject(widget_object);
    IupCommon.InitializeObject(widget_object);

    return handle_id;
  },
  emjsLabel_CreateSeparatorHorizontal: function(handle_id) {
    var widget_object = IupCommon.GetObjectForID(handle_id);
    var h = document.createElement("hr");
    widget_object.appendChild(h); 
  },

  emjsLabel_CreateSeparatorVertical: function(handle_id) {
    var widget_object = IupCommon.GetObjectForID(handle_id);
    alert("vertical separator");
  },


  emjsLabel_DestroyLabel: function(handle_id) {
    // Do I need to removeEventListener?
    IupCommon.DeleteObject(handle_id);
  },

  emjsLabel_SetTitle: function(handle_id, label_title) {
    // console.log("label title " + label_title);
    var widget_object = IupCommon.GetObjectForID(handle_id);
    if(widget_object)
    {
      widget_object.innerHTML = UTF8ToString(label_title);
    }
  },

  emjsLabel_SetPadding: function(handle_id, horiz, vert) {
    var widget_object = IupCommon.GetObjectForID(handle_id);

    // how does padding work in Iup?  Is it from corners or center?
    widget_object.style.paddingTop = vert + "px";
    widget_object.style.paddingBottom = vert + "px";
    widget_object.style.paddingLeft = horiz + "px";
    widget_object.style.paddingRight = horiz + "px";
  },

  emjsLabel_SetFGColor: function(handle_id, color) {
    var widget_object = IupCommon.GetObjectForID(handle_id);

    // what format is color? assuming its an rgb separated by spaces
    // console.log('handle id: ' + handle_id);
    // console.log('fg color: ' + color);
    // if (color) {
    //   var rgbVals = color.split(" ");
    //   var jsColor = 'rgb(' + rgbVals[0] + ', ' + rgbVals[1] + ', ' + rgbVals[2] + ')'; 
    //   widget_object.style.color = jsColor;
    // }
  },

  emjsLabel_SetBGColor: function(obj, val) {
    // console.log("bgcolor first arg: " + obj);
    // console.log("bgcolor sec arg: " + val);
  },

  emjsLabel_ToggleActive: function(handle_id, enable) {
    var widget_object = IupCommon.GetObjectForID(handle_id);
    enable === 1 ? widget_object.disabled = false : widget_object.disabled = true;
  },

  emjsLabel_SetAlignmentAttrib: function(handle_id, value) {
    // FUT - check to make sure ih->data->type != IUP_LABEL_SETP_HORIZ

    // get object to apply alignment to
    var widget_object = IupCommon.GetObjectForID(handle_id);
    var alignment = UTF8ToString(value);

    // works like ARIGHT:ABOTTOM
    if (alignment === "ARIGHT") {
      widget_object.style.height = "30px";
      widget_object.style.display = "table-cell";
      widget_object.style.verticalAlign = "middle";
      widget_object.style.textAlign = "right";  
    }
    else if (alignment === "ACENTER") {

      widget_object.style.height = "30px";
      widget_object.style.display = "table-cell";
      widget_object.style.verticalAlign = "middle";
      widget_object.style.textAlign = "center";
    }
    else {  // must be ALEFT

      widget_object.style.height = "30px";
      widget_object.style.display = "table-cell";
      widget_object.style.verticalAlign = "middle";
      widget_object.style.textAlign = "left"; 
    }

    // cant really align vertically in CSS like it does in gtk...
    // will need to further understand how the containers work and
    // maybe use padding or margin percentage to emulate

    return 1;
  },


  emjsLabel_SetImageAttrib: function(handle_id, pixel_ptr, array_length, width, height, pitch) {
//    var canvas_widget = IupCommon.GetObjectForID(handle_id);
	  // NOTE: Originally, I created a Canvas as the root object and just drew to the canvas.
	  // But growing/shrinking of the image as the canvas was resized does not appear to be built-in.
	  // But for the IMG object, we get resizing for free.
	  // The problem is how to go from a pixel array to an image.
	  // Turns out we can convert a Canvas to an Image using toDataURL.
	  // So I changed the root object to be IMG instead of CANVAS.
	  // This means I need to create a temporary CANVAS to render to an IMG.
	  // All the various data conversions we are doing concern me wrt to performance, but I'm not sure what can be done to improve this.
	  // I also don't know the impact of creating and collecting a temporary canvas object is.
	  // Nor do I know how much RAM keeping a temporary canvas in the background always resident would be.
	  // We could try to reuse/share a single canvas to cut down memory.
	  // However, if any of this stuff needs to be re-entrant (perhaps we start looking at async/threads?)
	  // then sharing a canvas would be a bad idea.
	  // (Also, for some of the IUP alignment attributes, it may turn out we need more control over the drawing and may need to revert back to Canvas.)
	  // So for now, we do the simple thing of creating a temporary canvas and collecting it.
	  var canvas_widget = document.createElement("CANVAS");
	  var canvas_context = canvas_widget.getContext("2d");


	  // We must set the Canvas size to match the image size or we get a final png that either has tons of extra padding space around the image or too small.
	  // TODO: I am not sure if I need to set both the canvas and context sizes.
	  canvas_widget.width  = width;
	  canvas_widget.height = height;
	  canvas_context.canvas.width  = width;
	  canvas_context.canvas.height = height;


	// pitch is effectively width (with alignment padding) * 4 bytes per pixel
    var pixal_data = Module.HEAPU8.slice(pixel_ptr, pixel_ptr + pitch * height);
//  var pixal_data = Module.HEAPU8.slice(pixel_ptr, pixel_ptr + width * height * 4);
    var image_data = canvas_context.getImageData(0, 0, width, height);
    image_data.data.set(pixal_data);
    canvas_context.putImageData(image_data, 0, 0);
	// At this point we have a usable Canvas with an image in it.

	// So now convert the Canvas into something we can use with IMG.
	var image_dataurl = canvas_widget.toDataURL("image/png");      
    var image_widget = IupCommon.GetObjectForID(handle_id);
	image_widget.src = image_dataurl;
  },

  emjsLabel_DropFilesTarget: function(handle_id) {
    var widget_object = IupCommon.GetObjectForID(handle_id);
    var elem = document.createElement('div');
    elem.style.border = '1px solid black';
    elem.style.height = '100px';
    elem.style.width = '200px';
    elem.ondrop = this.dropHandler(event);
    elem.ondragover = this.dragOverHandler(event);
  },

  dropHandler: function(ev) {
    console.log('File(s) dropped');

    // Prevent default behavior (Prevent file from being opened)
    ev.preventDefault();

    if (ev.dataTransfer.items) {
      // Use DataTransferItemList interface to access the file(s)
      for (var i = 0; i < ev.dataTransfer.items.length; i++) {
        // If dropped items aren't files, reject them
        if (ev.dataTransfer.items[i].kind === 'file') {
          var file = ev.dataTransfer.items[i].getAsFile();
          console.log('... file[' + i + '].name = ' + file.name);
        }
      }
    } else {
      // Use DataTransfer interface to access the file(s)
      for (var i = 0; i < ev.dataTransfer.files.length; i++) {
        console.log('... file[' + i + '].name = ' + ev.dataTransfer.files[i].name);
      }
    } 
    
    // Pass event to removeDragData for cleanup
    this.removeDragData(ev);
  },

  removeDragData: function(ev) {
    console.log('Removing drag data');

    if (ev.dataTransfer.items) {
      // Use DataTransferItemList interface to remove the drag data
      ev.dataTransfer.items.clear();
    } else {
      // Use DataTransfer interface to remove the drag data
      ev.dataTransfer.clearData();
    }
  },

  dragOverHandler: function(ev) {
    console.log('File(s) in drop zone'); 

    // Prevent default behavior (Prevent file from being opened)
    ev.preventDefault();
  },

  emjsLabel_EnableEllipsis: function(handle_id) {
    var widget_object = IupCommon.GetObjectForID(handle_id);
    widget_object.style.whiteSpace = "nowrap";
    widget_object.style.overflow = "hidden";
    widget_object.style.textOverflow = "ellipsis";
    // missing: -o-text-overflow: ellipsis;
  },

  emjsLabel_DisableWordWrap: function(handle_id) {
    var widget_object = IupCommon.GetObjectForID(handle_id);
    widget_object.style.whiteSpace = "-moz-pre-wrap"; /* Firefox */
    widget_object.style.whiteSpace = "-o-pre-wrap"; /* Opera */
    widget_object.style.whiteSpace = "pre-wrap"; /* Chrome */
    widget_object.style.wordWrap = "break-word"; /* IE */
  }

};

autoAddDeps(LibraryIupLabel, '$Label');
mergeInto(LibraryManager.library, LibraryIupLabel);

File: external/iup/src/emscripten/iupemscripten_label.c
/**
 * IupLabel module, C source
 * Please see accompanying JS file for external functions
 *
 * See Copyright Notice in "iup.h"
 */


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <memory.h>
#include <stdarg.h>

#include "iup.h"
#include "iupcbs.h"

#include "iup_object.h"
#include "iup_layout.h"
#include "iup_attrib.h"
#include "iup_str.h"
#include "iup_image.h"
#include "iup_label.h"
#include "iup_drv.h"
#include "iup_image.h"
#include "iup_focus.h"

#include "iup_childtree.h"

#include "iupemscripten_drv.h"

#include "iup_key.h"
#include <emscripten.h>
#include "SDL.h"

extern int emjsLabel_CreateLabel(void);
extern int emjsLabel_CreateImage(void);
extern int emjsLabel_CreateSeparatorHorizontal(void);
extern int emjsLabel_CreateSeparatorVertical(void);
extern void emjsLabel_SetTitle(int handle_id, const char* title);

extern void emjsLabel_CreateSeparator(int handle_id, char* type);
// this probably needs to return a char*
extern void emjsLabel_ToggleActive(int handle_id, int enable);
extern void emjsLabel_SetPadding(int handle_id, int horiz, int vert);
extern void emjsLabel_SetFGColor(int handle_id, const char* color); /* should it be constant char*? */
extern void emjsLabel_SetBGColor(int handle_id, char* color); 
extern int emjsLabel_SetAlignmentAttrib(int handle_id, const char* value);
extern void emjsLabel_EnableEllipsis(int handle_id);
extern void emjsLabel_DropFilesTarget(int handle_id);
extern void emjsLabel_SetImageAttrib(int handle_id, void* pixel_data, size_t array_length, int width, int height, int pitch);


// adds padding to element
void iupdrvLabelAddExtraPadding(Ihandle* ih, int *x, int *y)
{
  (void)ih;
  (void)x;
  (void)y;
}

static int emscriptenLabelSetTitleAttrib(Ihandle* ih, const char* value)
{
  // do we need this check? (it was in gtk)
  if (ih->data->type == IUP_LABEL_TEXT) {
    emjsLabel_SetTitle(ih->handle->handleID, value);
  }

	return 1;

}

static int emscriptenLabelSetPaddingAttrib(Ihandle* ih, const char* value)
{
  // it will try to set attributes before the widget is created - don't let it
  if (ih == NULL || ih->handle == NULL) {
    return 0;
  }
  else if (ih->handle->handleID == 0) {
    return 0;
  }
  else if (value == NULL) {
    value = "";
  }

  // split Iup's padding format (ex. "0x0") into ints
  iupStrToIntInt(value, &ih->data->horiz_padding, &ih->data->vert_padding, 'x');

  emjsLabel_SetPadding(ih->handle->handleID, ih->data->horiz_padding, ih->data->vert_padding);
  return 0;
}

static int emscriptenLabelSetFgColorAttrib(Ihandle* ih, const char* value)
{
  // call helper func from common; func will parse string and apply color to widget
  // can this be refactored into some other function? doesn't do much..
  iupdrvBaseSetFgColorAttrib(ih, value);

  return 1;
}

static int emscriptenLabelSetActiveAttrib(Ihandle* ih, const char* value)
{
  int enable = iupStrBoolean(value);

  /* update the inactive image if necessary */
  if (ih->data->type == IUP_LABEL_IMAGE)
  {
    if (!enable)
    {
      char* name = iupAttribGet(ih, "IMINACTIVE");
      if (name)
	  {
      //  gtkLabelSetPixbuf(ih, name, 0);
	  }
      else
      {
        /* if not defined then automatically create one based on IMAGE */
        name = iupAttribGet(ih, "IMAGE");
      //  gtkLabelSetPixbuf(ih, name, 1); /* make_inactive */
      }
    }
    else
    {
      /* must restore the normal image */
      char* name = iupAttribGet(ih, "IMAGE");
    //  gtkLabelSetPixbuf(ih, name, 0);
    }
  }

  //return iupBaseSetActiveAttrib(ih, value);
	

  emjsLabel_ToggleActive(ih->handle->handleID, enable);
  return 1; // ???
}

static int emscriptenLabelSetBgColorAttrib(Ihandle* ih, const char* value)
{
  emjsLabel_SetBGColor(7, "red");
  return 1;
}

static int emscriptenLabelSetWordWrapAttrib(Ihandle* ih, const char* value)
{
  return 0;
}

// TODO: check args for this function (its a callback)
// TODO: does this need to support multi-line text?
static int emscriptenLabelSetEllipsisAttrib(Ihandle* ih, const char* value)
{
  // assuming I dont need value right now (that is the string in the label being shortened)
  emjsLabel_EnableEllipsis(ih->handle->handleID);

	// if (ih->data->type == IUP_LABEL_TEXT) returns 1, all else returns 0 ?ERIC
  return 0;
}

static int emscriptenSetDropFilesTargetAttrib(Ihandle* ih, const char* value)
{
  emjsLabel_DropFilesTarget(ih->handle->handleID);
  // label_text or label_image returns 1, everything else returns 0? ?ERIC
  return 0; 
}

static int emscriptenLabelSetAlignmentAttrib(Ihandle* ih, const char* value)
{
  emjsLabel_SetAlignmentAttrib(ih->handle->handleID, value);

  // label_text or label_image returns 1, everything else returns 0? ?ERIC
  return 0;
}

/* { */

/*   GtkWidget* eventbox = (GtkWidget*)iupAttribGet(ih, "_IUP_EXTRAPARENT"); */
/*   unsigned char r, g, b; */

/*   /1* ignore given value, must use only from parent for the scrollbars *1/ */
/*   char* parent_value = iupBaseNativeParentGetBgColor(ih); */

/*   if (iupStrToRGB(parent_value, &r, &g, &b)) */
/*     iupgtkSetBgColor(eventbox, r, g, b); */

/*   (void)value; */
/*   return iupdrvBaseSetBgColorAttrib(ih, parent_value); */
/* } */


static int emscriptenLabelSetImageAttrib(Ihandle* ih, const char* value)
{
	
	if(ih->data->type == IUP_LABEL_IMAGE)
	{
		/*
		UIImageView* image_view = cocoaTouchLabelGetImageView(ih);
		if(NULL == image_view)
		{
			return 0;
		}
		*/
		char* name;
		int make_inactive = 0;
		
		if (iupdrvIsActive(ih))
		{
			make_inactive = 0;
		}
		else
		{
			name = iupAttribGet(ih, "IMINACTIVE");
			if (!name)
			{
				make_inactive = 1;
			}
		}
		
		
		// NOTE: Maybe SDL_surface is too low level, and we should be creating the JavaScript imageData.
		SDL_Surface* sdl_surface;
		sdl_surface = iupImageGetImage(value, ih, make_inactive, NULL);
		int width;
		int height;
		int bpp;
		
		iupdrvImageGetInfo(sdl_surface, &width, &height, &bpp);
		
		int pitch = sdl_surface->pitch;
		size_t array_length = pitch * height;
iupEmscripten_Log("emjsLabel_SetImageAttrib id:%d, ptr:0x%p, arr_len:%d, w:%d, h:%d, p:%d", ih->handle->handleID, sdl_surface->pixels, array_length, width, height, pitch);
		
		emjsLabel_SetImageAttrib(ih->handle->handleID, sdl_surface->pixels, array_length, width, height, pitch);

		
/*
		// FIXME: What if the width and height change? Do we change it or leave it alone?
		CGSize new_size = CGSizeMake(width, height);
		CGRect the_frame = [image_view frame];
		the_frame.size = new_size;
		[image_view setFrame:the_frame];

		[image_view setImage:the_bitmap];
*/		
		return 1;
	}
	else
	{
		return 0;
	}
}

static int emscriptenLabelMapMethod(Ihandle* ih)
{
  int label_id = 0;
  InativeHandle* new_handle = NULL;


  // Set text inside label (uses title attribute)
#if 0
	char* attrib_title = iupAttribGet(ih, "TITLE");
	if(attrib_title && *attrib_title != 0)
	{
		ih->data->type |= IUP_LABEL_TEXT;
		emjsLabel_SetTitle(label_id, attrib_title);
	}
#endif
	
	char* value = iupAttribGet(ih, "SEPARATOR");
	if (value)
	{
		if (iupStrEqualNoCase(value, "HORIZONTAL"))
		{
			ih->data->type = IUP_LABEL_SEP_HORIZ;

			label_id = emjsLabel_CreateSeparatorHorizontal();

			
		}
		else /* "VERTICAL" */
		{
			ih->data->type = IUP_LABEL_SEP_VERT;
			label_id = emjsLabel_CreateSeparatorVertical();


		}

		
	}
	else
	{

		value = iupAttribGet(ih, "IMAGE");
		if (value)
		{
			ih->data->type = IUP_LABEL_IMAGE;
			/*
			char *name;
			int make_inactive = 0;
			
			if (iupdrvIsActive(ih))
			{
				name = iupAttribGet(ih, "IMAGE");
			}
			else
			{
				name = iupAttribGet(ih, "IMINACTIVE");
				if (!name)
				{
					name = iupAttribGet(ih, "IMAGE");
					make_inactive = 1;
				}
			}
			*/

			label_id = emjsLabel_CreateImage();
			
		}
		else
		{
			ih->data->type = IUP_LABEL_TEXT;


			label_id = emjsLabel_CreateLabel();
	

		
		}
	}

  new_handle = (InativeHandle*)calloc(1, sizeof(InativeHandle));

  new_handle->handleID = label_id;
  ih->handle = new_handle;

  iupEmscripten_SetIntKeyForIhandleValue(label_id, ih);


	iupEmscripten_AddWidgetToParent(ih);
	
	
	/* configure for DRAG&DROP of files */
	if (IupGetCallback(ih, "DROPFILES_CB"))
	{
		iupAttribSet(ih, "DROPFILESTARGET", "YES");
	}
	return IUP_NOERROR;
}


static void emscriptenLabelUnMapMethod(Ihandle* ih)
{
	/*
	id the_label = ih->handle;
	[the_label release];
	ih->handle = nil;
	*/

}

void iupdrvLabelInitClass(Iclass* ic)
{
  /* Driver Dependent Class functions */
  ic->Map = emscriptenLabelMapMethod;
//	ic->UnMap = emscriptenLabelUnMapMethod;


  /* Driver Dependent Attribute functions */
  iupClassRegisterAttribute(ic, "ACTIVE", iupBaseGetActiveAttrib, emscriptenLabelSetActiveAttrib, IUPAF_SAMEASSYSTEM, "YES", IUPAF_DEFAULT);

  iupClassRegisterAttribute(ic, "BGCOLOR", iupBaseNativeParentGetBgColorAttrib, emscriptenLabelSetBgColorAttrib, IUPAF_SAMEASSYSTEM, "DLGBGCOLOR", IUPAF_DEFAULT);

  /* iupClassRegisterAttribute(ic, "FGCOLOR", NULL, iupdrvBaseSetFgColorAttrib, IUPAF_SAMEASSYSTEM, "DLGFGCOLOR", IUPAF_DEFAULT); */

  iupClassRegisterAttribute(ic, "FGCOLOR", NULL, emscriptenLabelSetFgColorAttrib, IUPAF_SAMEASSYSTEM, "DLGFGCOLOR", IUPAF_DEFAULT);
  iupClassRegisterAttribute(ic, "PADDING", iupLabelGetPaddingAttrib, emscriptenLabelSetPaddingAttrib, IUPAF_SAMEASSYSTEM, "0x0", IUPAF_NOT_MAPPED);
  iupClassRegisterAttribute(ic, "TITLE", NULL, emscriptenLabelSetTitleAttrib, NULL, NULL, IUPAF_NO_DEFAULTVALUE|IUPAF_NO_INHERIT);
  iupClassRegisterAttribute(ic, "WORDWRAP", NULL, emscriptenLabelSetWordWrapAttrib, NULL, NULL, IUPAF_DEFAULT);

  // ERIC?: how does alignment work?
  iupClassRegisterAttribute(ic, "ALIGNMENT", NULL, emscriptenLabelSetAlignmentAttrib, "ALEFT:ACENTER", NULL, IUPAF_NO_INHERIT);  /* force new default value */

  iupClassRegisterAttribute(ic, "ELLIPSIS", NULL, emscriptenLabelSetEllipsisAttrib, NULL, NULL, IUPAF_DEFAULT);

  iupClassRegisterAttribute(ic, "DROPFILESTARGET", NULL, emscriptenSetDropFilesTargetAttrib, NULL, NULL, IUPAF_DEFAULT);
  // gtk defined in iup_dragdrop.c with IUPAF_NO_INHERIT as the last larg - which should be used?


  // ERIC?:
  // dropfilestarget
  // turn div/widget into area that files can be dropped onto to trigger a file upload dialog; as simple as possible
  
  /* IupLabel only */
  // upon compile, image is compiled into binary format, possibly need special apis to read
  // emscripten should provide these apis
  iupClassRegisterAttribute(ic, "IMAGE", NULL, emscriptenLabelSetImageAttrib, NULL, NULL, IUPAF_IHANDLENAME|IUPAF_NO_DEFAULTVALUE|IUPAF_NO_INHERIT);
#if 0

  /* IupLabel GTK and Motif only */
  iupClassRegisterAttribute(ic, "IMINACTIVE", NULL, gtkLabelSetImInactiveAttrib, NULL, NULL, IUPAF_IHANDLENAME|IUPAF_NO_DEFAULTVALUE|IUPAF_NO_INHERIT);

  /* IupLabel GTK only */
  iupClassRegisterAttribute(ic, "MARKUP", NULL, NULL, NULL, NULL, IUPAF_DEFAULT);
#endif
}
